adding additional_secure_json_data

rockaut opened the pull request:
##### SUMMARY

fixes #106 
fixes #107
fixes #109 
fixes #110 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
grafana_datasource

##### ADDITIONAL INFORMATION

**IMPORTANT: this PR might change previous behaviour!** because:
- As until now (#107) once a datasource where created or updated with any secureJsonData further changes wouldn't be updated and reporting changed=false. With this PR each run will change the fields, which should be the expected result anyways!
- Ad #110: Users don't HAVE to change anything as from now on password isn't save to secureJsonData.password but just password. Runs won't report change=True like above. BUT: they have to switch to secure_password if they want to use encrypted passwords.

rockaut commented:
Question guys @rrey @seuf WDYT:
With this changes as stated the task with secureJsonData related stuff is always reporting changed=true.
IMO I think this is ok, yes even what I want as user as this is transparent - currently it is implemented in this way and pointed out in the docs.

Alternative:
```python

    if state == 'present':
        payload = get_datasource_payload(module.params)
        if ds is None:
            result = grafana_iface.create_datasource(payload)
            ds = grafana_iface.datasource_by_name(name)
            module.exit_json(changed=True, datasource=ds, msg='Datasource %s created' % name)
        else:
            diff = compare_datasources(payload.copy(), ds.copy())
            if diff.get('before') == diff.get('after'):
                module.exit_json(changed=False, datasource=ds, msg='Datasource %s unchanged' % name)
            grafana_iface.update_datasource(ds.get('id'), payload)
            ds = grafana_iface.datasource_by_name(name)

```

Currently the diff comes from before update_datasource and handles if it should be updated. To circumvent above changed=True thematic we could again diff AFTER the update with the secureJsonData/secureJsonFields portions poped and just report the other changes. This way it would work again in old behavior.
Also we could take a split and add a `dontReportSecureDataChanges: true` property and go booth ways. Thought this might be confusing.

IMO:
- implementing `dontReportSecureDataChanges` with default to `true` for now while pointing it out noticeable for all users
- after some releases changeing it to `false` per default and marking it as `deprecated` for the next release
- after that removing this special shit and move on :D

rockaut commented:
I've now implemented `dontReportSecureDataChanges` as module property as described above with defaulting to `dontReportSecureDataChanges=True`. I see it as a backward compat mode for now.

Also as pointed out in #111 I've changed all version_added to 1.0.0 for the release.

seuf commented:
Actually, the current behavior is to report `changed=True` when `secureJsonFields` is present. So the default of the new parameter `dontReportSecureDataChanges` should be `false`. 
I've tested with tls_client_cert and tls_client_key parameters on the elasticsearch datasource integration test and it reports a change :
```
TASK [grafana_datasource : Check elasticsearch datasource creation idempotency] *************************************************************************************************************************************************************************************************
--- before
+++ after
@@ -17,11 +17,6 @@
     "name": "datasource-elastic",
     "orgId": 1,
     "password": "",
-    "secureJsonFields": {
-        "tlsCACert": true,
-        "tlsClientCert": true,
-        "tlsClientKey": true
-    },
     "type": "elasticsearch",
     "url": "https://elastic.company.com:9200",
     "user": "",

changed: [testhost]
```

rockaut commented:
@seuf  the current behaviour is inconsistent if anything:
- it removes secureJsonFields as a whole on changes with tls*
- it removes secureJsonData as a whole in any case

So your test might be correct for this specific example yes because it only incorporates tls*.

```python
def compare_datasources(new, current):
    del current['typeLogoUrl']
    del current['id']
    if 'version' in current:
        del current['version']
    if 'readOnly' in current:
        del current['readOnly']
    if current['basicAuth'] is False:
        del current['basicAuthUser']
        del current['basicAuthPassword']
    if 'jsonData' in current:
        if 'tlsAuth' in current['jsonData'] and current['jsonData']['tlsAuth'] is False:
            del current['secureJsonFields']
        if 'tlsAuth' not in current['jsonData']:
            del current['secureJsonFields']
    # secureJsonData contains ciphered data which we cannot compare
    if 'secureJsonData' in new:
        del new['secureJsonData']
    return dict(before=current, after=new)
```

Check the cloudwatch tests for example:
Create reports changed=True while the next create reports changed=False which shouldn't be true because the accessKey is secureJsonData and should be changed every run. Which it isn't because #107 and only because this it's changed=False.

```
TASK [grafana_datasource : Create cloudwatch datasource] ***********************
changed: [testhost] => {"changed": true, "datasource": {"access": "proxy", "basicAuth": false, "basicAuthPassword": "", "basicAuthUser": "", "database": "", "id": 4, "isDefault": false, "jsonData": {"authType": "keys", "customMetricsNamespaces": "n1,n2", "defaultRegion": "us-west-1", "tlsAuth": false, "tlsAuthWithCACert": false}, "name": "datasource-cloudwatch", "orgId": 1, "password": "", "readOnly": false, "secureJsonFields": {"accessKey": true, "secretKey": true}, "type": "cloudwatch", "typeLogoUrl": "", "url": "http://monitoring.us-west-1.amazonaws.com", "user": "", "version": 1, "withCredentials": false}, "msg": "Datasource datasource-cloudwatch created"}

TASK [grafana_datasource : Check cloudwatch datasource creation idempotency] ***
ok: [testhost] => {"changed": false, "datasource": {"access": "proxy", "basicAuth": false, "basicAuthPassword": "", "basicAuthUser": "", "database": "", "id": 4, "isDefault": false, "jsonData": {"authType": "keys", "customMetricsNamespaces": "n1,n2", "defaultRegion": "us-west-1", "tlsAuth": false, "tlsAuthWithCACert": false}, "name": "datasource-cloudwatch", "orgId": 1, "password": "", "readOnly": false, "secureJsonFields": {"accessKey": true, "secretKey": true}, "type": "cloudwatch", "typeLogoUrl": "", "url": "http://monitoring.us-west-1.amazonaws.com", "user": "", "version": 1, "withCredentials": false}, "msg": "Datasource datasource-cloudwatch unchanged (secureJson ignored!)"}
```

rrey commented:
Hi !
Thanks @rockaut for your work around this.

The fix described in the PR is ok for me, I like the alternative too but the dontReportSecureDataChanges usage bothers me  because using it will hide changes actually performed by the module. I would prefer that we don’t introduce it but I don’t want to have final word here, I like the discussion around this.

rockaut commented:
You're right imo, I don't like it either tbt. I've implemented it just so the module will be backwards compatible.

So I guess you and @seuf basically think the same. Then the easiest thing would be to set the property to false by default. I would also mark it as deprecated so that with the next major it can get removed.

Diff should already be fine as I've already implemented it with respect to this.

That said I will have to change the integration tests as most of them assume the old behavior.

rrey commented:
I’m on vacation but I’ll try to review your patch when possible

rockaut commented:
I changed the default to false as discussed.

rockaut commented:
So that I get it right:
 - so instead of the `dontReportSecureDataChanges` you mean `enforce_secure_data`
 - and of course fixing the problem that secure data wasn't ever updated
 - but also keeping the `additional_secure_json_data`
 - but DONT update as long enforce_secure_data is set?

So let's bring it in an workflow:
- One user sets the secure_password on his pg datasource
- on first creation the secure data is being written and changed=True
- on subsequent updates it's changed=False because secure data IS NOT updated
- if a user want the secure data changed one has to set `enforce_secure_data: true`
- if so subsequent updates are changed=True everytime

If you mean it like this it's basically just changing the property name and make it False by default.

Hmmm... I think for now ( Version 1.0.0 ) this might be good. But it still doesn't solves each Question/Behaviour imo.
We at least have to note it on each secure property and in the notes. The flow might not be clear. But I kind of understand what you mean. I also don't like it when modules do make changes but don't report them - it's just wrong.

I will change it.

rockaut commented on plugins/modules/grafana_datasource.py:
```diff
@@ -587,13 +632,15 @@ def main():
         database=dict(type='str', default=""),
         user=dict(default='', type='str'),
         password=dict(default='', no_log=True, type='str'),
+        secure_password=dict(default='', no_log=True, type='str'),
```
Yes it could. In this case I just saw that a dedicated property might be a good idea in conjunction to `password:` to make it clear the later one is unencrypted while the new one is.

rockaut commented on plugins/modules/grafana_datasource.py:
```diff
@@ -279,9 +288,29 @@
 extends_documentation_fragment:
 - community.grafana.basic_auth
 - community.grafana.api_key
+notes:
```
Hmm... see my comments down below. I will rewrite it accordingly to the new behavior.

rockaut commented:
@rrey you were right. I like this way better!
I changed it to the described way and updated the notes and docs. I also added two examples highlighting how to handle the enforce_secure_data in plays.
For tests I currently have added it in elastics tests. I will add it to other tests later maybe in another PR so each tests has it for safety and completeness.

rockaut commented on plugins/modules/grafana_datasource.py:
```diff
@@ -407,14 +470,26 @@ def compare_datasources(new, current):
-        del new['secureJsonData']
+
+    # check if secureJsonData should be compared
+    if not compareSecureData:
```
No it can't. If you compare the data without dropping the data it will always has a diff and thus will always get updated and reporting changed=True.

rockaut replied:
If you leave that out the whole enforce_secure_data doesn't work. Except of course you want to handle it elsewhere, but i suppose it's not what you mean?

rockaut replied:
Also if you don't handle the possibility that the secureJsonData might be empty you can again run into diff even if there is no diff.

rockaut commented on plugins/modules/grafana_datasource.py:
```diff
@@ -65,6 +67,12 @@
     - The datasource password.
+    - For encrypted password use C(secure_password).
+    type: str
+  secure_password:
```
:( but you're the maintainer. I thought it would be a more user-friendly way.
I will change it after we settled in the enforce_secure_data conversation.

rockaut replied:
Can we at least agree on a note in the option description for password? #113 proves that there is not only me getting confused.

rockaut commented:
@rrey should I remove the secure_password as you advised?

rrey commented:
If you have time to do it and check that it won’t break anything else, it would be awesome
